Cautious (#33)
siemens/OOASP: asp_interactive_configuration/DOC.md

1. *Complete:* Getting a complete configuration extending `C`.
2. *Check:* Checking `C` for errors.

> 3. *Options:* Obtaining all possible classes, values, and associations to complete `C`.

Reviewer: For 3., I find the old version of the text better understandable. If it is still correct, I suggest reverting back to it.

**Externals**

- **`user(ooasp_isa(CLASS,ID))`**
  The user assigned class `CLASS` to object `ID`

- **`user(ooasp_associated(ASSOC,ID1,ID2))`**
  The user associated `ID1` and `ID2` via `ASSOC`

- **`user(ooasp_attribute_value(N,ID,VALUE))`**
  The user selected value `VALUE` for attribute `N` of object `ID`

The truth value of these externals is defined by the [Interactive Configurator](#interactive-configurator) based on the current (partial) configuration that is being constructed using an object of class [OOASPConfiguration](#configuration).

> The user externals are generated restricted by the possible classes that an object can have. This is defined by the argument `cls` passed in the grounding, which corresponds to the domain defined by `ooasp_domain(_,cls).`. Therefore, if the new object is grounded for a class `cls`, only valid attributes and associations for `cls` and its subclasses will be considered. This is summarized in the auxiliary predicate `counts_as(O,C)`, which states that object `O` can count as class `C`.

Reviewer: `ooasp_domain` is mentioned nowhere else in DOC.md, which makes this explanation a bit hard to understand. Furthermore, looking at usages of this predicate in the encoding, the class seems to be the first argument, not the second. Is `ooasp_domain(_,cls)` correct then?

Author: I added a phrase above, corrected the order and rephrased this part.

## Numerical attributes

Numerical attributes are treated using the system `fclingo`. This systems adds a founded condition while using `clorm` in the background. 
Attributes whose type is `int`, such as `nr_passengers`, `standing_room`, and `nr_seats` in the Metro example, are treated as variables with this system as explained below. Note that numerical values can also be treated as an enumeration using the type `enumint` instead. This option will consider all values in the defined range for predicate `ooasp_attr_value`, such as `frame_position` in the racks example.

### Usage

> When an attribute `ATTR` is of type `int`, their value is defined using an `fclingo` variable `ooasp_attr_fvalue(ATTR,O)`.  The value of such variable will be associated to the value of attribute `ATTR` for object `O`.

Reviewer: ```suggestion
When an attribute `ATTR` is of type `int`, their value is defined using an `fclingo` variable `ooasp_attr_fvalue(ATTR,O)`.  The value of such a variable will be associated to the value of attribute `ATTR` for object `O`.
```

> These values can be compared using the theory symbol `&fsum`. For instance, in the following rule we use `&fsum{ooasp_attr_fvalue(A,new_object)}<MIN` to state that the value of `A` is smaller than `MIN` leading to a constraint violation. Notice that the variable `MIN` must be defined in a positive literal of the body.

Reviewer: ```suggestion
These values can be compared using the theory symbol `&fsum`. For instance, in the following rule we use `&fsum{ooasp_attr_fvalue(A,new_object)}<MIN` to state that the value of `A` being smaller than `MIN` will lead to a constraint violation. Notice that the variable `MIN` must be defined in a positive literal of the body.
```

```
ooasp_cv(value_outside_of_range,new_object,"Value for {} outside min range {}",(A,MIN)) :-
	 ooasp_attr(C,A,int),
	 ooasp_isa(C,new_object),
	 &fsum{ooasp_attr_fvalue(A,new_object)}<MIN,
	 ooasp_attr_minInclusive(C,A,MIN).
```

To check if a value is defined one can do so by checking if it is equal to itself. In the case of the next rule, if no value is defined, then the constraint violation is inferred.

```
ooasp_cv(no_value,new_object,"Missing value for {}",(A,)) :-
	ooasp_attr(C,A,int),
	ooasp_isa(C,new_object),
	not &fsum{ooasp_attr_fvalue(A,new_object)}=ooasp_attr_fvalue(A,new_object).
```

> This type of attributes must be treated differently to the rest, therefore they require rules like to ones above to handle each case. Below we show the difference of writing a constraint violation using `ooasp_attr_value/3` and `ooasp_attr_fvalue/2`. Notice that in the second case, we directly compare the values by summing them up.

Reviewer: ```suggestion
This type of attributes must be treated differently to the rest, therefore they require rules like the ones above to handle each case. Below we show the difference of writing a constraint violation using `ooasp_attr_value/3` and `ooasp_attr_fvalue/2`. Notice that in the second case, we directly compare the values by summing them up.
```

    ooasp_attr_value(nr_seats,new_object,NRS),
    ooasp_attr_value(standing_room,new_object,S),
    NRP!=NRS+S.
```

#### `ooasp_attr_fvalue/2`
```
ooasp_cv(nrpassengers_sum,new_object,"Number of passengers not adding up",(new_object,)):-
    ooasp_isa(wagon,new_object),
    &fsum{ooasp_attr_fvalue(nr_seats,new_object);
          ooasp_attr_fvalue(standing_room,new_object)}!=
        ooasp_attr_fvalue(nr_passengers,new_object).
```

### Interactivity

The assigned values for these attributes are handled internally by the `fclingo `system and are only accessed after the models are found. Therefore, they are not considered in any brave or cautious consequences. This means that we do not know in advance the possible values an attribute can take for a partial configuration. Interactively, this can be solved in different ways, such as using a slider instead of a dropdown or some other numerical input. For convenience, we keep the usage in the jupyter notebooks with dropdowns and add all possible values for the attribute, even though they might create an invalid configuration. Invalid configurations can still be checked and fixed in the UI. 

> In this version, externals are used for every possible value to allow user input. However, this is not efficient. Another version is proposed in branch `efficient-fclingo` where externals are added on demand. However. This option does not allow for the task `check`, since it requires the choices to be active. 

Reviewer: ```suggestion
In this version, externals are used for every possible value to allow user input. However, this is not efficient. Another version is proposed in branch `efficient-fclingo` where externals are added on demand. However, this option does not allow for the task `check`, since it requires the choices to be active. 
```

```prolog

> #program domain(new_object).

Reviewer: Shouldn't this be `#program domain(new_object, cls).`?

Consider the following rule:

```prolog

> arity(new_object,ASSOC,ARITY):- ARITY = #count{ID2:ooasp_assoc(ASSOC,new_object,_,_,ID2,_,_)}.

Reviewer: ```suggestion
arity(new_object,ASSOC,ARITY) :- ARITY = #count{ID2:ooasp_assoc(ASSOC,new_object,_,_,ID2,_,_)}.
```

	ooasp_assoc_limit(ASSOC,max,OPT,C,CMAX),
	ooasp_isa(C,ID1),
	#count { ID2:ooasp_associated_general(ASSOC,OPT,ID1,ID2) } > CMAX,
	active(new_object).
```

Notice that, as before, this rule must consider the `active(new_object)` to define in which step the constraint is violated. The count, however, does not depend on the `new_object`.

We use here two auxiliary predicates to avoid having to repeat the rule:

- **`ooasp_assoc_limit(ASSOC,max,POS,C,CMAX,C2)`**
  The association `ASSOC` has a maximum limit of `CMAX` for objects of class `C` at `POS`, where `POS` is `1` or `2` and refers to one end of the association (1: left end, 2: right end), where the other side of the associations are objects of class `C2`.
  *Example: `ooasp_assoc_limit(ass,min,1,racks,10)`:
	In the association `ass`, each `racks` object must be associated to at least `10` elements*

- **`ooasp_associated_general(ASSOC,POS,ID1,ID2)`**
  Object `ID1` appearing in position `POS` of association `ASSOC` is associated to object `ID2`
  *Example: `ooasp_associated_general(c1,ass,2,23,10)`: object `23` of the second class of association `ass` is associated with object `10`*

--- 

## Incremental solving

> This multi-shot encoding can directly be used for incremental solving. With this idea we ground for the current objects and try to find a configuration using Task 1 *Complete*, if there is no complete configuration (returns UNSAT) with this number of objects, we ground for one more object and repeat the process until we have a satisfiable answer.

Reviewer: ```suggestion
This multi-shot encoding can directly be used for incremental solving. With this idea we ground for the current objects and try to find a configuration using Task 1 *Complete*, and if there is no complete configuration (returns UNSAT) with this number of objects, we ground for one more object and repeat the process until we have a satisfiable answer.
```

> This incremental process find the smallest configuration. However, it has the drawback that right before getting the satisfiable answer the solving time increases since it has to look over the whole search space, which impacts in performance.

Reviewer: ```suggestion
This incremental process finds the smallest configuration. However, it has the drawback that right before getting the satisfiable answer the solving time increases since the solver has to look over the whole search space, which impacts its performance.
```

> Additionally, if there are cases where no configuration can be found, the incremental solving will continue adding objects util the set limit is reached. For instance, in the metro example, if one Wagon is added with number of passengers 0, and a seat object is added, then this partial configuration can not reach a compleat configuration, since it would require two wagons, which is not possible. However, the incremental solving would not know this and continue looking for the answer.

Reviewer: ```suggestion
Additionally, if there are cases where no configuration can be found, the incremental solving will continue adding objects util the set limit is reached. For instance, in the metro example, if one Wagon is added with number of passengers 0, and a seat object is added, then this partial configuration can not reach a complete configuration, since it would require two wagons, which is not possible. However, the incremental solving would not know this and continue looking for the answer.
```

###  Create required

> Creates all the objects required by the current objects. To do so, we use the cautious consequences of the program for getting the constraint violations that are common to all models, and proceed to solve one of them at a time. For this process we need optimization statements, to get rid of models that include unnecessary  constraint. This optimization is only activate for getting cautious consequences. 

Reviewer: What are models that "include unnecessary constraint[s]"?

Author: I rephrased, the details and examples are in the subsections below

> Creates all the objects required by the current objects. To do so, we use the cautious consequences of the program for getting the constraint violations that are common to all models, and proceed to solve one of them at a time. For this process we need optimization statements, to get rid of models that include unnecessary  constraint. This optimization is only activate for getting cautious consequences. 

Reviewer: ```suggestion
Creates all the objects required by the current objects. To do so, we use the cautious consequences of the program for getting the constraint violations that are common to all models, and proceed to solve one of them at a time. For this process we need optimization statements, to get rid of models that include unnecessary constraint. This optimization is only active for getting cautious consequences. 
```

> #### Optimize number of lower bond violations

Reviewer: ```suggestion
#### Optimize number of lower bound violations
```

> The main optimization with priority `3`, penalizes lower bound violations based on how many objects have not yet been associated `CMIN-N`. This week constraint will prioritize models that associate as many objects as possible. For instance, take a partial configuration with one rack and 3 frames. Since the external `ooasp_potential_cv` is false we would get a model for associating 0,1,2 and 3 frames. We this week constraint we will only get one where 3 frames are associated and the lower bound is only violated with one missing frame for the `rack_frames` association.

Reviewer: ```suggestion
The main optimization with priority `3` penalizes lower bound violations based on how many objects have not yet been associated `CMIN-N`. This weak constraint will prioritize models that associate as many objects as possible. For instance, take a partial configuration with one rack and 3 frames. Since the external `ooasp_potential_cv` is false we would get a model for associating 0,1,2 and 3 frames. With this weak constraint we will only get one where 3 frames are associated and the lower bound is only violated with one missing frame for the `rack_frames` association.
```

```
:~ ooasp_cv(lowerbound,ID,_,(ASSOC,CMIN,N,C,OPT,new_object)). [CMIN-N@3,(ASSOC,ID)]
```

#### Optimize associations to smallest objects

This optimization can be viewed like a symmetry constraint. The idea is to try to associate the smallest objects first by penalizing based on the ids of the objects associated. This is done in two constraints to consider both positions for the association. With this in place, we would make sure that if we have one rack and 5 frames, always the frame with the largest id would remain not associated, those making the same constraint violation appear in all stable models and consequently in the cautious consequences. Without this optimization, we would have a model for each frame being left out and no constraint violation would appear in the intersection. 

#### Using the cautions consequences to create required objects

1. Gets the cautious consequences for the program where potential cv can be violated and the optimization is activated.

> 2. Takes the first lower bound constraint violation. This predicate already has the information of how many objects where associated and what the minimum was. Based on this information it crates and associates the required objects of the specific class

Reviewer: ```suggestion
2. Takes the first lower bound constraint violation. This predicate already has the information of how many objects were associated and what the minimum was. Based on this information it creates and associates the required objects of the specific class.
```

Notice that this process is not so efficient when we have multiple constraint violations that add a single object. Such is the case in the racks example, when we have many Elements, each Element generates a constraint violation which adds one module at a time, leading to multiple solve calls. 

### Smart incremental solving 

In the smart incremental solving the performance issue is tackled by adding inferences in every iteration.

This is done using the create required task in three steps:
- Try to get a solution for the current number of objects, 
- If this is UNSAT then call create required
- If no object is added with the create required, then proceed as in normal incremental by adding one object of type `object``

## Advanced features to simplify writing domain specific constraint violations

### Association specialization

Association specializations will specialize the lower and upper bound of an association for subclasses.

For example, the association `rack_frames` between class `rack` and `frame` is defined by the atom `ooasp_assoc(rack_frames,rack,1,1,frame,4,8)`.

> The racks example limits the number of racks associated to the racks subclass `rackSingle` to be exactly `4`. This is achieved by adding an association specialization with name `rack_framesS` which specializes `rack_frames`, by the atom `ooasp_assoc_specialization(rack_framesS,rack_frames)`. This means that all `rack_frameS` associations are also `rack_frames` associations. Then, the upper and lowe bound can be redefined when the association is defined in `ooasp_assoc(rack_framesS,rackSingle,0,1,frame,4,4)`. 

Reviewer: ```suggestion
The racks example limits the number of racks associated to the racks subclass `rackSingle` to be exactly `4`. This is achieved by adding an association specialization with name `rack_framesS` which specializes `rack_frames`, by the atom `ooasp_assoc_specialization(rack_framesS,rack_frames)`. This means that all `rack_framesS` associations are also `rack_frames` associations. Then, the upper and lower bound can be redefined when the association is defined in `ooasp_assoc(rack_framesS,rackSingle,0,1,frame,4,4)`. 
```

While upper bounds can be restricted, lower bounds are relaxed. So that the lower bounds of an association specialization are smaller or equal to the original one.
Notice that in our example the lower bound of how many racks singles are associated to a frame is now `0`, otherwise frames would be forced to be associated to a `rackSingle`. The requirement of having some rack associated comes from the super association `rack_frames`. 

### Specialized constraint violations

Specialized constraint violations are those used in different instances but not encoded in the knowledge base. Some of them are the following:

- **(unique)** all attribute values of objects in an association are different 
- **(same)** all attribute values of objects in an association are equal (exceptions)
- **(table)** which value combinations are allowed
- **(specialization)** attribute value determines the type of objects

> These constraints can be defined in the knowledge base using the predicate `ooasp_special_cv/2`, where the first argument is the name of the constraint and the second one has any additional information needed. To implement a specialized constraint violation, the file [ooasp/encodings/ooa_checksp.lp](ooasp/encodings_check/ooasp.lp) must be extended with the needed constraints.

Reviewer: ```suggestion
### Special constraint violations

Special constraint violations are those used in different instances but not encoded in the knowledge base. Some of them are the following:

- **(unique)** all attribute values of objects in an association are different 
- **(same)** all attribute values of objects in an association are equal (exceptions)
- **(table)** which value combinations are allowed
- **(specialization)** attribute value determines the type of objects

These constraints can be defined in the knowledge base using the predicate `ooasp_special_cv/2`, where the first argument is the name of the constraint and the second one has any additional information needed. To implement a special constraint violation, the file [ooasp/encodings/ooa_checksp.lp](ooasp/encodings_check/ooasp.lp) must be extended with the needed constraints.
```
I propose to rename "specialized constraint violations" to "special constraint violations" to avoid confusion since, especially having read the section directly preceding this section, the reader could wrongly assume a conceptual connection to association specialization.

> This process was done for the  **unique** constraint violation. It is used in the knowladge base of the racks example, to make sure that the frame positions of all frames associated to a rack are different. In this case the second argument is a tuple `(ASSOCIATION, CLASS, ATTRIBUTE)`

Reviewer: ```suggestion
This process was done for the  **unique** constraint violation. It is used in the knowledge base of the racks example, to make sure that the frame positions of all frames associated to a rack are different. In this case the second argument is a tuple `(ASSOCIATION, CLASS, ATTRIBUTE)`
```
